packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx

```diff
@@ -19,84 +20,161 @@ export interface IDetailsRowFieldsProps {
+      oldColumns: { ...{}, ...this.props.columns }, // get by value instead of by reference
+      oldCellContent: DetailsRowFields._getCellContent(this.props),
+      oldItem: { ...{}, ...this.props.item }
+    });
   }
 
   public render(): JSX.Element {
-    const { columns, columnStartIndex, shimmer } = this.props;
-    const { cellContent } = this.state;
+    const { item, columns } = this.props;
+    const { cellContent, renderingPhase, oldColumns, oldCellContent, oldItem } = this.state;
+
+    /* When a column gets new data, we animate the transition. The old data slides out
+     * and then the new data slides in. In order to do this, we remember the old data
+     * whenever new props come in, in getDerivedStateFromProps(). Then we do rendering
+     * in three phases:
+     * 2: Render with the old data and animate it sliding out
+     * 1: Render the new data and slide it in
+     * 0: resting state, remove all animation classes
+     * The phases go 2->1->0 with a delay inbetween each to allow the animation to
+     * complete. Each cell will also check to make sure it changed, if its contents
+     * did not change, then it will not do any animation.
+    */
+
+    if (renderingPhase) {
+      this._async.setTimeout(() => {
+        this.setState({ renderingPhase: renderingPhase - 1 });
+      }, Number(AnimationVariables.durationValue4) * 1000 /* convert to ms */);
+    }
 
     return (
       <div
         className={css('ms-DetailsRow-fields', styles.fields)}
         data-automationid="DetailsRowFields"
         role="presentation"
       >
-        {columns.map((column, columnIndex) => (
-          <div
-            key={columnIndex}
-            role={column.isRowHeader ? 'rowheader' : 'gridcell'}
-            aria-colindex={columnIndex + columnStartIndex + 1}
-            className={css(
-              'ms-DetailsRow-cell',
-              styles.cell,
-              column.className,
-              column.isMultiline && 'is-multiline',
-              column.isRowHeader && styles.isRowHeader,
-              column.isPadded && styles.isPadded,
-              column.isMultiline && styles.isMultiline,
-              column.isIconOnly && shimmer && styles.shimmerIconPlaceholder,
-              shimmer && styles.shimmer
-            )}
-            style={{ width: column.calculatedWidth! + INNER_PADDING + (column.isPadded ? ISPADDED_WIDTH : 0) }}
-            data-automationid="DetailsRowCell"
-            data-automation-key={column.key}
-          >
-            {cellContent[columnIndex]}
-          </div>
-        ))}
+        {columns.map((column, columnIndex) => {
+          const oldColumn = oldColumns && oldColumns[columnIndex];
+          let contentChanged = false;
+          if (oldItem && oldColumn) {
+            contentChanged =
+              DetailsRowFields._getCellText(item, column) !== DetailsRowFields._getCellText(oldItem, oldColumn);
+          }
+          switch (renderingPhase) {
```

Comment:
Make this an actual enum for clarity. Use const enum so it only adds the numbers though. As long as you assign numbers to the enums you can treat them like numbers.

Reply:
@joschect I used numbers so I can easily do the renderingPhase-- in the logic above. Since it's internal state, it shouldn't need to be exposed/documented anyways.

Reply:
You can just set them as numbers and then you can use them as numbers. It makes things more immediately obvious to the next person, even if it's documented.